Allow user to change editor font at runtime.

cben opened the pull request:
This probably needs more polish and testing, but won't have more time for a couple days so sending to review as-is to see what you think.
It Works For Me™ but I shuffled a lot of code around, there must be bugs.

Refactored settings <-> QFont round tripping.  I tried several ways
(including a second layer of __setattr__ magic) and think this is best.

Dropped special-casing of preview font being unset - I'm not sure if the
special behavior was useful but anyway there is no UI to unset the font
once you've set it.
However with empty conf file, this is causing
```
QFont::setPointSize: Point size <= 0 (0), must be greater than 0
```
warnings, and they repeat a lot — I think the 0/actual mismatch is defeating my QFontSetting caching.

I'm not sure `'font': 'Sans'` is a sane default.  It exists on Linux with typical fontconfig.
Similarly for `'monospace'` — not sure what it'd do on other platforms.
I lifted `Sans` from a place where it was going into CSS, but CSS would want `sans-serif`.
To get the precise "system's default sans/monospace font" effect with QFont it seems we'd need a "style hint".

We're discarding all font info except family and size.  If user chooses a demi-bold cursive variant, why not allow it?  The only reason I see is it might make `*italic*` and/or `**bold**` syntax highlight invisble.
QFont docs recommends just storing its toString in a single QSettings value.

mitya57 commented:
Thanks for the pull request, I will look at it a bit later.

Please do not submit more pull requests on GitHub until I officially move the project from SourceForge to here.

mitya57 commented on ReText/window.py:
```diff
@@ -22,7 +23,7 @@
 from PyQt5.QtCore import QDir, QFile, QFileInfo, QFileSystemWatcher, \
  QIODevice, QLocale, QRect, QTextCodec, QTextStream, QTimer, QUrl, Qt
-from PyQt5.QtGui import QColor, QDesktopServices, QFont, QFontMetrics, QIcon, \
+from PyQt5.QtGui import QColor, QDesktopServices, QFont, QIcon, \
```
`QFont` import is no longer needed.

mitya57 commented on ReText/window.py:
```diff
@@ -9,7 +9,8 @@
 from markups.common import MODULE_HOME_PAGE
 from ReText import icon_path, DOCTYPE_MARKDOWN, DOCTYPE_REST, app_name, \
  app_version, globalSettings, settings, readListFromSettings, \
```
`settings` import is no longer needed.

mitya57 commented:
Thanks! I left some notes, but apparently I left half of them on the commit, not on the diff (and these are different things on GitHub).

If you fix all of them, I will merge it (preferably amend the existing commit).

cben commented:
I'm seeing that export File > Export > PDF uses the preview font but File > Export > HTML, File > Export > ODT, Edit -> View Html code don't use it.
Is that expected or a new bug?

Thanks for the review.
Addressed your comments.
`if size:` indeed fixed the warning.
I've checked with some debugging prints => the QFont caching in get() only works when font & size are set in config.  But get() is only called for some user actions - toggling preview, export - so I don't think it's a problem.

Squashed to single commit. [original commits: 228978165b5b3b548686a588fcfce67d771051c4, 398841931e4474ea44cdf4dd508707ab8fa4ba8b, 319c563eb2c4e62bd3c8318e8fcde886b42f84b2]

BTW, all my testing was on Ubuntu.  I could test on windows 10 if you want.

> Please do not submit more pull requests on GitHub until I officially move the project from SourceForge to here.

OK.

mitya57 commented:
Please test on Windows 10, we need to make sure that `Sans` as font name works.

mitya57 commented on ReText/__init__.py:
```diff
@@ -128,10 +130,34 @@ def __setattr__(self, option, value):
+		font = self.cached
+		if not font or font.family() != family or font.pointSize() != size:
+			self.cached = font = QFont()
+			font.setFamily(family)
```
Btw, what do you think about wrapping this in `if family:`?

This will allow you to not hardcode `Sans` and let Qt figure out what font to use itself. (In this case the special case for CSS will need to be restored).

mitya57 commented:
The new official ReText repository is https://github.com/retext-project/retext. I will delete this one once we finish the discussion on this pull request.

cben commented:
Good news: install on windows 10 works, can even be simplified a bit.

- windows 10 home 64 bit
- The PyQt5 page has no installers for python 3.5; I wondered whether I can  get it another way, perhaps just try "pip install".  But https://stackoverflow.com/questions/32582797/python3-5-on-win7x64-can-not-load-from-qt5-module-qtwidgets says that won't work, so I'm sticking to Python 3.4.
- python 3.4.3 64 bit.
- PyQt5-5.5-gpl-Py3.4-Qt5.5.0-x64.exe	 (Full install)

Now I figured I don't want to permanently set env vars if I can avoid it.
Also, by now python should come with a working pip out of the box, so skipped the ez_setup.

- C:\Python34\Scripts\pip.exe install Pygments
- C:\Python34\Scripts\pip.exe install Markdown
- C:\Python34\Scripts\pip.exe install docutils
- C:\Python34\Scripts\pip.exe install Markups
- cd C:\Users\beni\Documents\GitHub\retext
- C:\Python34\python.exe retext.py

Success!

![2015-09-16 15 49 20](https://cloud.githubusercontent.com/assets/273688/9907931/11049860-5c9a-11e5-9835-fe1384d47f43.png)

(Except there are no toolbar icons.
I thought I'm being clever by running from a git checkout instead of downloading retext and downloading icons.
Turns out git source has an icons/ dir but that's only 2 files, I still needed the icons .tgz on top of it.  That fixed it.)

Bad news: monospace and Sans defaults don't work - even without my changes.
The screenshot above was with `master` checked out and it's using a proportional font on the left and some serif font on the right.

OK news: with my change, both font choosers work on windows:
![2015-09-16 16 18 51](https://cloud.githubusercontent.com/assets/273688/9908129/ead72378-5c9a-11e5-8596-4a558116826a.png)
(that's after choosing Comic Sans and Impact IIRC)

[Unrelated news:
As the example text mentions in the above screenshots, enabling math via Preferences didn't work.  But after creating the .config directory manually, it did work.  Though it's very ugly:
![2015-09-16 16 25 11](https://cloud.githubusercontent.com/assets/273688/9908201/3d01372e-5c9b-11e5-9cc9-c7a32e82a102.png)
that's also after unpacking the toolbar icons.  BTW, it seems the migration to github didn't include the toolbar icons, they only exist as a tarball download on sourceforge.
]

cben commented:
I don't currently have more time to debug on windows.
I suspect the right solution does require separate Qt / CSS code paths for default case.

mitya57 commented:
I have refactored the code working with fonts, and added the editor font choosing based on this pull request. This work is in the new repository (https://github.com/retext-project/retext), I am deleting this one now. Please open new issues or pull requests at the new location.
